druid/src/text/layout.rs

```diff
@@ -0,0 +1,257 @@
+    font: KeyOrValue<FontDescriptor>,
+    text_size_override: Option<KeyOrValue<f64>>,
+    text_color: KeyOrValue<Color>,
+    //FIXME: all this caching stuff can go away when we have a simple way of
+    // checking if something has changed in the env.
+    cached_text_color: Color,
+    cached_font: FontDescriptor,
+    // when set, this will be used to override the size in
+    // the font descriptor. This provides an easy way to
+    // change only the font size, while still using a `FontDescriptor` in the `Env`.
+    cached_text_size: Option<f64>,
+    // the underlying layout object. This is constructed lazily.
+    layout: Option<PietTextLayout>,
+}
+
+impl TextLayout {
+    /// Create a new `TextLayout` object.
+    ///
+    /// You do not provide the actual text at creation time; instead you pass
+    /// it in when calling [`rebuild_if_needed`].
+    ///
+    /// [`rebuild_if_needed`]: #method.rebuild_if_needed
+    //FIXME: we may change to having this take the text initially, but this
+    //depends on us having some clone-friendly text types.
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        TextLayout {
+            font: crate::theme::UI_FONT.into(),
+            cached_font: Default::default(),
+            text_color: crate::theme::LABEL_COLOR.into(),
+            cached_text_color: Color::BLACK,
+            text_size_override: None,
+            cached_text_size: None,
+            layout: None,
+        }
+    }
+
+    /// Returns `true` if this layout needs to be rebuilt.
+    ///
+    /// This happens (for instance) after style attributes are modified.
+    ///
+    /// This does not account for things like the text changing, handling that
+    /// is the responsibility of the user.
+    pub fn needs_rebuild(&self) -> bool {
+        self.layout.is_none()
+    }
+
+    /// Set the default text color for this layout.
+    pub fn set_text_color(&mut self, color: impl Into<KeyOrValue<Color>>) {
+        self.text_color = color.into();
+        self.layout = None;
+    }
+
+    /// Set the default font.
+    ///
+    /// The argument is a [`FontDescriptor`] or a [`Key<FontDescriptor>`] that
+    /// can be resolved from the [`Env`].
+    ///
+    /// [`FontDescriptor`]: struct.FontDescriptor.html
+    /// [`Env`]: struct.Env.html
+    /// [`Key<FontDescriptor>`]: struct.Key.html
+    pub fn set_font(&mut self, font: impl Into<KeyOrValue<FontDescriptor>>) {
+        self.font = font.into();
+        self.layout = None;
+        self.text_size_override = None;
+    }
+
+    /// Set the font size.
+    ///
+    /// This overrides the size in the [`FontDescriptor`] provided to [`set_font`].
+    ///
+    /// [`set_font`]: #method.set_font.html
+    /// [`FontDescriptor`]: struct.FontDescriptor.html
+    pub fn set_text_size(&mut self, size: impl Into<KeyOrValue<f64>>) {
+        self.text_size_override = Some(size.into());
+        self.layout = None;
+    }
+
+    /// The size of the laid-out text.
```

Comment:
Maybe should mention that it's only meaningful after `rebuild_if_needed` has been called?